### PR TITLE
Feature footprint collision check and coverage update

### PR DESCRIPTION
**Pull request by AronTie98**
Besides the already reviewed visualization changes, this feature branch focuses on adding the capability to compute footprints. To be more specific, the cells in a grid covered by this footprint. It uses these footprints (of either the robot or its cleaning tool) to check if a manoeuvres (which is essentially a concatenation of footprints) is feasible. Currently, standard manoeuvres are turning in place (left/right) and making a step forward, or only making a step forward. Feasibility is determined using collision as a hard constraint and already visited grids as a soft constraint (a maximum number is allowed). Based on that, it builds the collision-free spirals while it marks the grid cells covered by the tool every step as visited. 

Screenshot from 2021-11-11 14-02-47

--> Note: only one spiral is plotted here, as the A* planning between grids doesn't work with footprints/orientations

### CODE REVIEW

**Comment by AronTie98** [comment]
Another example of one spiral but now with division factor 7 (and overlap for turns = 9)

image

Both in this example and the previous one, the tool footprint is set as { {0.53, 0.53}, {0.53, -0.53}, {-0.53, -0.53}, {-0.53, 0.53} }

**Comment by AronTie98** [comment]
Additionally, the spiraling towards a wall is shown in the figure below.

image

--> Note: the gap above the starting position will need to be filled in with another spiral.

File: include/full_coverage_path_planner/spiral_stc.hpp [hunk 1]
```diff
@@ -23,22 +23,43 @@
...
 
 using namespace std::chrono_literals;
 using std::string;
+
 namespace full_coverage_path_planner
 {
   class SpiralSTC : public nav2_core::GlobalPlanner , private full_coverage_path_planner::FullCoveragePathPlanner
   {
   public:
+
```
Comment (cesar-lopez-mar) [inline include/full_coverage_path_planner/spiral_stc.hpp:36]: are these variables parameters? if so, please add a comment and group them in a block. Also, internal variables are declared under the protected label of the class (You do not want a user to access them from outside the class)

File: src/spiral_stc.cpp [hunk 2]
```diff
@@ -90,64 +101,193 @@ namespace full_coverage_path_planner
...
+    {
+      it--; // Let iterator point to second to last element
+    }
     gridNode_t prev = *(it);
+
+    // Initialize spiral direction towards robot's y-axis
+    int dx = 0;
+    int dy = 1;
+    int dx_prev;
+    double yaw1 = yaw_start; // This value will change later on
```
Comment (cesar-lopez-mar) [inline src/spiral_stc.cpp:120]: please use a more descriptive name

File: src/spiral_stc.cpp [hunk 3]
```diff
@@ -90,64 +101,193 @@ namespace full_coverage_path_planner
...
       }
-      // This condition might change in the loop bellow
-      done = true;
-      // loop over the four possible directions: up, right, down, left
-      for (size_t i = 0; i < 4; ++i)
+      done = true; // This condition might change in the loop below
+      // Loop over the three possible directions: left, forward, right (directions taken before counter-clockwise turn)
+      for (size_t i = 0; i < 3; ++i)
       {
+        int x1 = pathNodes.back().pos.x;
```
Comment (cesar-lopez-mar) [inline src/spiral_stc.cpp:151]: Also in these variables please use more descriptive names ( some of them were already there )

Comment (nxdefiant) [reply src/spiral_stc.cpp:151]: I find 'd' for distance descriptive..

File: include/full_coverage_path_planner/spiral_stc.hpp (hunk 1 above)
Comment (AronTie98) [reply include/full_coverage_path_planner/spiral_stc.hpp]: Indeed, I have added the ones that will definitely stay to the node parameters and wrote an explanation for them in the README file in the latest commit. The remaining debug and testing parameters are grouped and moved under the protected label of the class.

File: src/spiral_stc.cpp (hunk 2 above)
Comment (AronTie98) [reply src/spiral_stc.cpp]: I changed it (and others) in the latest commit.

File: src/spiral_stc.cpp (hunk 3 above)
Comment (AronTie98) [reply src/spiral_stc.cpp]: See the latest commit.

**Comment by AronTie98** [comment]
Much has changed in the meantime, in both functionality and code style. The new feature branch will open a pull request in the near future.

**Pull request closed without merging**